File: evals/turn-table.ts
import { mkdirSync, writeFileSync } from 'node:fs'
import { dirname } from 'node:path'
import { readJsonLines } from './lib/jsonl.ts'
import { fromExportedEvent, groupTurns } from './lib/records.ts'
import { projectTurn, toCsv } from './lib/turn-table.ts'

/**
 * Raw traces in, one publishable row per turn out. The projection is an allowlist, so this is
 * the boundary where private data stops: `evals/data/` is gitignored, `evals/results/` is not.
 *
 * Records from before the envelope existed (2026-08-06) carry no `turnId` and are dropped rather
 * than reconstructed — mixing two schemas into one table is how a denominator goes wrong quietly.
 */
const IN = 'evals/data/traces.jsonl'
const OUT = 'evals/results/turns.csv'

const raw = readJsonLines<unknown>(IN, { required: true })
const usable = raw.map(fromExportedEvent).filter((r) => r !== null)
const turns = groupTurns(usable)
const rows = turns.map(projectTurn)

mkdirSync(dirname(OUT), { recursive: true })
writeFileSync(OUT, `${toCsv(rows)}\n`)

const bySource = new Map<string, number>()
for (const row of rows) {
  bySource.set(String(row.source), (bySource.get(String(row.source)) ?? 0) + 1)
}

console.log(
  `${raw.length} records in, ${usable.length} with an envelope (${raw.length - usable.length} pre-schema, dropped)`,
)
console.log(`${turns.length} turns → ${OUT}`)
for (const [source, n] of [...bySource].sort((a, b) => a[0].localeCompare(b[0]))) {
  console.log(`  ${String(n).padStart(4)}  ${source}`)
}
const unfinished = rows.filter((r) => r.outcome === 'unfinished').length
if (unfinished) {
  console.log(`  ${String(unfinished).padStart(4)}  unfinished (no turn record)`)
}

File: evals/turn_table.py
from collections import Counter

from lib.jsonl import read_json_lines
from lib.paths import DATA, RESULTS
from lib.records import from_exported_event, group_turns
from lib.turn_table import project_turn, to_csv

IN = DATA / 'traces.jsonl'
OUT = RESULTS / 'turns.csv'


def main() -> None:
    raw = read_json_lines(IN, required=True)
    usable = [r for r in (from_exported_event(event) for event in raw) if r is not None]
    turns = group_turns(usable)
    rows = [project_turn(t) for t in turns]

    OUT.parent.mkdir(parents=True, exist_ok=True)
    OUT.write_text(to_csv(rows) + '\n', encoding='utf-8')

    by_source = Counter(str(row['source']) for row in rows)
    dropped = len(raw) - len(usable)
    print(f'{len(raw)} records in, {len(usable)} with an envelope ({dropped} pre-schema, dropped)')
    print(f'{len(turns)} turns → {OUT}')
    for source, n in sorted(by_source.items()):
        print(f'  {n:>4}  {source}')
    unfinished = sum(1 for row in rows if row['outcome'] == 'unfinished')
    if unfinished:
        print(f'  {unfinished:>4}  unfinished (no turn record)')


if __name__ == '__main__':
    main()
